### PR TITLE
Add project specific nexus user

### PR DESCRIPTION
**Pull request by tjaeschke**

### CODE REVIEW

**Comment by tjaeschke** [comment]
relates to https://github.com/opendevstack/ods-quickstarters/issues/58

**Comment by tjaeschke** [comment]
PR ist still work in progress, but if you want to tke a look, this would be great.

File: infrastructure-setup/scripts/prepare-nexus.sh [hunk 1]
```diff
@@ -3,18 +3,31 @@ export PATH=$PATH:/usr/local/bin/
...
+  . $BASE_DIR/local.env.config
+fi
+
+
 oc login -u system:admin
 oc project cd
 
 cd /ods/ods-core/nexus/ocp-config
-yes 'y' | tailor update -v --force
+# yes 'y' | tailor update -v --force
```
Comment (michaelsauter) [inline infrastructure-setup/scripts/prepare-nexus.sh]: Why is it commented now? BTW, there is an `--non-interactive` flag ...

File: create-projects/Jenkinsfile [hunk 2]
```diff
@@ -67,8 +70,15 @@ podTemplate(
       sh "sh ./create-projects/create-projects.sh"
     }
 
-    stage('creating jenkins') {
-      sh "sh ./create-projects/create-cd-jenkins.sh"
+    // read json reult from Nexus curl call
+    def nexusResponse = readJSON text: sh(script: "curl -sf --user $nexusUser:$nexusPass -X POST -H 'Accept: application/json' -H 'Content-Type: text/plain' \"$nexusHost/service/rest/v1/script/createProjectAccess/run\" -d $projectId", returnStdout: true).trim()
+
+    def specificUser = readJSON text: nexusResponse.result
```
Comment (michaelsauter) [inline create-projects/Jenkinsfile]: Why `readJSON` twice? Shouldn't nexusResponse be a JSON object already?

File: docs/modules/nexus/pages/index.adoc [hunk 3]
```diff
@@ -1,2 +1,40 @@
...
+== Setup
+
+The OpenShift templates are located in `ocp-config` and can be compared with the OC cluster using https://github.com/opendevstack/tailor[tailor]. For example, run `cd ocp-config && tailor status` to see if there is any drift between current and desired state.
+
+You are also able to use the script `prepare-nexus.sh` located in the `infrastructure-setup/scripts` folder of this repository.
+You will have to customize the script to match your target environment.
+
+If you have a running Nexus 3 instance and you only want to add scripted resources, take a look at the `create-nexus-resources.sh` script. +
+Inside this script you will find curl calls, that are used to run scripts in the Nexus 3 scripting API. +
+The called scripts are located in the `json` subfolder and can be customized by your need. +
```
Comment (michaelsauter) [inline docs/modules/nexus/pages/index.adoc]: There seem to be some stray `+` signs at the end of the lines?

File: docs/modules/nexus/pages/index.adoc [hunk 4]
```diff
@@ -1,2 +1,40 @@
...
+*Please keep in mind, that the user, who is provided in the curl calls has to have script execution rights in Nexus 3!*
+
+The `createProjectSpecificAccess.json` contains the logic to create a project specific nexus user to isolate project artifacts from each other. +
+The logic uses content selectors with an expression that uses the `path` to identify repos, that belong to one project, so you will have to specify a naming convention, if you use private repositories and isolated projects.
```
Comment (michaelsauter) [inline docs/modules/nexus/pages/index.adoc]: I'm not sure I am getting this. Is the CSEL now a path within a repo, or a repo? I thought the former, but this paragraph makes me unsure now.

File: docs/modules/nexus/pages/index.adoc [hunk 5]
```diff
@@ -1,2 +1,40 @@
...
+
+You can modify the used expression in the `createProjectSpecificAccess.json` script if it doesn't match your requirements and upload it to Nexus 3 with the related curl call.
```
Comment (michaelsauter) [inline docs/modules/nexus/pages/index.adoc]: I think we need an example for the default expression to understand what this means, e.g. for Maven and NPM repos.

File: infrastructure-setup/scripts/prepare-nexus.sh (hunk 1 above)
Comment (tjaeschke) [reply infrastructure-setup/scripts/prepare-nexus.sh]: That happened yesterday during local testing and wasn't intended. I have corrected this.

File: create-projects/Jenkinsfile (hunk 2 above)
Comment (tjaeschke) [reply create-projects/Jenkinsfile]: I had the same assumption, but unfortunately it isn't. After reading the nexusResponse into JSON the result field is still a string and has to be converted to an JSON object for itself.

File: docs/modules/nexus/pages/index.adoc (hunk 3 above)
Comment (tjaeschke) [reply docs/modules/nexus/pages/index.adoc]: The `+` is intended. It marks a linebreak in ASCIIDOC

File: docs/modules/nexus/pages/index.adoc (hunk 4 above)
Comment (tjaeschke) [reply docs/modules/nexus/pages/index.adoc]: CSEL is the expression language used for a content selector and you are right, it is used to retrieve artifacts from an existing repository matching the expression. In the provided implementation the content selector is set to match all repositories, so even older and new repositories will match.

File: docs/modules/nexus/pages/index.adoc (hunk 5 above)
Comment (tjaeschke) [reply docs/modules/nexus/pages/index.adoc]: I have added an explanation

**Comment by segator** [comment]
Questions:
* contect selector depends of repo type (npm, mvn , pip..) how you plan to do that,
* when you apply the contect selector per project, per project creation(to all repo types?) when quickstarter request?(depends of the quickstarter type create one or other content selector?)

**Comment by tjaeschke** [comment]
> Questions:
> 
> * contect selector depends of repo type (npm, mvn , pip..) how you plan to do that,
> * when you apply the contect selector per project, per project creation(to all repo types?) when quickstarter request?(depends of the quickstarter type create one or other content selector?)

The current base implementation targets the path as minimal common ground, so the selector targets all existing and all future repos or registries in Nexus containing the path pattern.

**Comment by clemensutschig** [comment]
@michaelsauter - what are we doing with this now?

**Comment by michaelsauter** [comment]
Hmm, I guess right now it is out of scope for v3. I suggest to revisit after 3.0 if we want to prioritise this for v4.

**Comment by clemensutschig** [comment]
@michaelsauter inlude.. or close ;-)?

**Comment by michaelsauter** [comment]
Let's close. When the topic comes up again, we can still refer to the work done here.

**Pull request closed without merging**